feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go

```diff
@@ -0,0 +1,625 @@
+}
+
+// Function to read a table entry returns response/error
+func readTableEntry(args *testArgs, t *testing.T) (*p4_v1.ReadResponse, error) {
+	t.Helper()
+	readMsg := &p4_v1.ReadRequest{
+		DeviceId: args.deviceID,
+		Entities: []*p4_v1.Entity{
+			{
+				Entity: &p4_v1.Entity_TableEntry{
+					TableEntry: &p4_v1.TableEntry{
+						TableId: p4rtutils.WbbTableMap["acl_wbb_ingress_table"],
+					},
+				},
+			},
+		},
+	}
+	readClient, err := args.handle.Read(readMsg)
+	if err != nil {
+		t.Logf("ReadClient Error: %v", err)
+		return nil, err
+	}
+	readResp, err := readClient.Recv()
+	if err != nil {
+		t.Logf("readResp Error: %v", err)
+		return nil, err
+	}
+	return readResp, nil
+}
+
+// Every client/controller should be able to read
+func canRead(t *testing.T, args *testArgs) (bool, error) {
+	_, readErr := readTableEntry(args, t)
+	if readErr != nil {
+		return false, readErr
+	}
+	return true, nil
+}
+
+// Only Primary should be able to write
+func canWrite(t *testing.T, args *testArgs) (bool, error) {
+	pktIO := getGDPParameter(t)
+	writeErr := writeTableEntry(args, t, pktIO, false)
+	if writeErr != nil {
+		if args.wantWrite == false {
+			t.Logf("Expected P4Client error: Client write permission (highID %d, lowID %d): want %v, got %v", args.highID, args.lowID, args.wantWrite, false)
+		}
+		return false, writeErr
+	}
+	return true, nil
+}
+
+// This function handles the values passed through the test case
+// and compares them to values obtained in canRead()/canWrite().
+func validateRWResp(t *testing.T, args *testArgs) bool {
+	returnReadVal := true
+	returnWriteVal := true
+	// Validate write permissions
+	writeOp, writeErr := canWrite(t, args)
+	if args.wantWrite != writeOp {
+		t.Errorf("Client write permission error: (highID %d, lowID %d): want %v, got %v", args.highID, args.lowID, args.wantWrite, writeOp)
+		if writeErr != nil {
+			t.Errorf("Write Error %v", writeErr)
+		}
+		returnWriteVal = false
+	}
+
+	// Validate read permissions
+	readOp, readErr := canRead(t, args)
+	if args.wantRead != readOp {
+		t.Errorf("Client read permission error: (highID %d, lowID %d): want %v, got %v", args.highID, args.lowID, args.wantRead, readOp)
+		t.Errorf("Read Error: %v", readErr)
+		returnReadVal = false
+	}
+	if returnWriteVal && returnReadVal {
+		return true
+	}
+	return false
+}
+
+func removeClient(handle *p4rt_client.P4RTClient) {
+	handle.StreamChannelDestroy(&streamName)
+	handle.ServerDisconnect()
+}
+
+// Test Zero client with 0 electionID
+func TestZeroMaster(t *testing.T) {
+	dut := ondatra.DUT(t, "dut")
+	configureDeviceId(t, dut)
+	configurePortId(t, dut)
+	test := testArgs{
+		desc:       pDesc,
+		lowID:      inId0,
+		highID:     inId0,
+		handle:     clientConnection(t, dut),
+		deviceID:   deviceId,
+		wantFail:   true,
+		wantStatus: 0,
+	}
+	t.Run(test.desc, func(t *testing.T) {
+		resp, err := streamP4RTArb(&test)
+		if err == nil && test.wantFail {
+			t.Errorf("Zero ElectionID (0,0) is allowed: %v", err)
```

Comment:
Is this a requirement? Couldn't find anything related to this here: https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-arbitration-updates

Reply:
There was some debate on this, however the lack of clarification of the spec is what prompted for a decision. @robshakir, thoughts?  I checked on a few vendors and they considered (0,0) as invalid election ID as well.

The closest thing I found in the spec under `16.2. Client Arbitration Update` was: 
> The sender need not specify an election_id. If the election_id is not specified, the sender's election_id is considered lower than any election_id, and the sender will thus never become primary. This way, a controller can choose to be a backup controller, in order to avoid “flapping” (if a backup controller connects to the switch shortly before the actual primary controller, therefore becoming primary temporarily).

Reply:
In master Arbitration, election ID zero need no be treated as a special case. The P4RT server should be able to accept the connection and client with election ID 0 may become primary, if its the only client getting connected.

Reply:
@rsindhujha, IIRC there was a conversation with @robshakir where we had concluded that zero election ID was not to become primary even if it was the only client getting connected.  Perhaps Rob can share thoughts here.

Reply:
This looks somewhat relevant: https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-zero-as-reserved-value